[ONNX] Fix scalar_type_analysis metadata for copied constant

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #86648
* __->__ #86716

Fix the source of metadata for copied constant. Since the constant is being implicitly casted,
it makes more sense to assign code location and etc with the user node.
This issue was discovered in https://github.com/pytorch/pytorch/issues/86627. This PR also adds unit test coverage for scope
information of nodes when they are altered by CSE and related passes.

BowenBao committed:
[ONNX] Fix scalar_type_analysis metadata for copied constant

[ghstack-poisoned]

justinchuby commented on test/onnx/test_utility_funs.py:
```diff
@@ -1097,6 +1097,110 @@ def forward(self, x, y, z):
+        #       so we expect 3 constants with different scopes. The 3 constants are for the 3 layers.
+        #       If CSE in exporter is improved later, this test needs to be updated.
+        #       It should expect 1 constant, with same scope as root.
+        SCOPE_PREFIX = "test_utility_funs.TestUtilityFuns_opset9.test_scope_of_constants_when_combined_by_cse_pass.<locals>"
```
nit: snake case. UPPER reserved for global constants

justinchuby reviewed (commented):
just a nit

justinchuby commented on torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp:
```diff
@@ -292,7 +292,7 @@ static void UpdateScalarTypeForInputs(
         const_node->t_(attr::value, new_val);
         const_node->insertBefore(n);
         const_node->output()->setType(TensorType::create(new_val));
-        const_node->copyMetadata(input->node());
```
Maybe a comment here to hint future readers?

justinchuby replied:
This lines looks the same as 291. Just for my learning - why do we do it twice?

AllenTiTaiWang replied:
+1 copyMetadata is a nightmare to me. I see them everywhere and not exactly know what they contribute ...

BowenBao replied:
> +1 copyMetadata is a nightmare to me. I see them everywhere and not exactly know what they contribute ...

lol I often need to check its source code to confirm. The name is too general. Probably a good idea to add some comments. 

> This lines looks the same as 291. Just for my learning - why do we do it twice?

My guess is unfound merge 'conflict'.

BowenBao committed:
Update on "[ONNX] Fix scalar_type_analysis metadata for copied constant"


Fix the source of metadata for copied constant. Since the constant is being implicitly casted,
it makes more sense to assign code location and etc with the user node.
This issue was discovered in https://github.com/pytorch/pytorch/issues/86627. This PR also adds unit test coverage for scope
information of nodes when they are altered by CSE and related passes.


[ghstack-poisoned]

BowenBao commented on torch/csrc/jit/ir/ir.h:
```diff
@@ -423,6 +423,7 @@ struct TORCH_API Node {
     return scope_->namesFromRoot();
   }
 
+  // Copies the source range, scope and callstack from another node.
```
@kit1980 I added a comment here outside of onnx merge rules, please help review this PR.

thiagocrepaldi commented on test/onnx/test_utility_funs.py:
```diff
@@ -1097,6 +1097,110 @@ def forward(self, x, y, z):
+
+        # NOTE: Duplicated constants are populated due to implicit casting in scalar_type_analysis,
+        #       so we expect 3 constants with different scopes. The 3 constants are for the 3 layers.
+        #       If CSE in exporter is improved later, this test needs to be updated.
```
That means that even if current implementation (CSE enabled as a jit pass for exporter) the constant will be duplicated 3 times instead of being folded to a single constant?

BowenBao replied:
Correct. The constants were created by scalar_type_analysis pass. This PR only corrects the meta data of those constants.

thiagocrepaldi replied:
Why implicit cast is needed, leading to duplicates? would it be preventable?

BowenBao replied:
torch supports implicit cast for arithmetic operators and more. ONNX does not. Hence exporter needs to do the casting. 
The duplication can only be prevented by examining the source and uses of the constants, otherwise it wouldn't be safe to delete the original constant, as it might still be used elsewhere. The process is essentially the CSE pass.

thiagocrepaldi reviewed (approved):
LGTM

malfet commented:
@pytorchbot merge
